CIPs/cip-granda_mento.md

```diff
@@ -0,0 +1,186 @@
+
+## Motivation
+
+There are no existing avenues that are able to satisfy high volume stable token exchanges on the order of millions. Users looking to make larger exchanges involving stable tokens are faced with 3 options:
+1. Exchange `CELO <-> stable token` via Mento slowly over a longer period of time.
+2. Arrange one or many `XXX <-> stable token` OTC trades.
+3. Create a large limit order of `fiat <-> stable token` on a centralized exchange in hope it gets filled over time.
+
+The existing implementation of Mento is suitable for facilitating low volume exchanges, e.g. up to thousands of stable tokens at a time, with minimal slippage. For cUSD, slippage of over 2% starts to occur on both Mento and on centralized exchanges at trade sizes of ~$50k+. OTC trading satisfies the needs of medium volume exchanges, e.g. ~$100k, but similarly takes a hit of around 2-3%. For lower total supply stable tokens with lower liquidity, larger OTC trades or centralized exchange orders may not be possible.
+
+Increasing Mento bucket sizes permanently to allow larger exchanges via Mento will reduce slippage, but it puts a more significant portion of the Reserve at risk.
+
+Building a process to exchange large (~$1 million+) amounts of CELO for stable tokens will enable large entities to make significant block purchases of stable tokens at scale. While minting of the stable token is expected to be the main use case, reversibility (ie exchanging `stable token -> CELO`) is desirable to instill higher confidence when exchanging large quantities.
+
+While Granda Mento will help facilitate `CELO <-> stable token` exchanges, this is ultimately meant as a mechanism to help purchasers with fiat (e.g. USD) who are seeking to own Celo's corresponding stable token (e.g. cUSD). Purchasers would come to an agreement with a broker (e.g. the Celo Foundation) who owns an existing amount of CELO. The broker would be the one to make the `CELO -> cUSD` exchange on-chain.
+
+The proposed implementation is being considered in the medium-term for large scale exchanges until Mento is possibly reworked to support more efficient large-scale minting on a sustained basis.
+
+## Specification
+
+At a high level, the design involves:
+1. Anyone can submit an exchange proposal.
+   * The amount of stable tokens being bought/sold in the exchange is required to be within a governable range.
+   * The assets being sold (ie either the stable token or CELO) in the exchange are deposited.
+   * The current oracle price for the exchange is recorded.
```

Comment:
Although this is significantly more complicated to implement on-chain, would it be more accurate to take a multi-day price average? This is often how large trades are priced OTC.

Reply:
This was discussed, ultimately because of the human "counterparties" that make up the approver multisig & governance, it was deemed that the approver/governance could disagree with the oracle price if they want, and that it's not worth the implementation effort & risk of implementing an on-chain TWAP. I feel like it could also be argued that just a straight up TWAP isn't a totally "fair" price, eg if the CELO price just jumped (which it has done in the past) then the TWAP may be low and the exchanger is trying their best to get a better deal than they would've if the normal oracle price was used
I'll beef up the `Rationale` section with this information, wdyt of the decision?

Reply:
It really makes sense, as the protocol can reject if the pricing goes really bad. Regarding the TWAP being "too good" in case of a price jump, that is really not a problem due liquidity constraints, for example: 
* 90 days TWAP is $3
* CELO is currently trading at $3.5
* Someone is making a proposal to buy 1M celo @TWAP = 3 cUSD

Although that scenario seems unfair, in reality the purchaser would in fact struggle to get a price of $3 if it wanted to liquidate that whole position on a CEX. I wouldn't personally push to veto a trade of that kind.

Maybe something we could do is allow the exchanger to submit their own price along with the exchange proposal, as long as it is within a certain range from the oracle price?

Without this "price discovery" mechanism, it could be that parties are reluctant to use Granda Mento in times of high volatility, or try to time the submission to get a good price, or even manipulate the price.

Reply:
I think a TWAP method would discourage anyone from using it during long periods of time when the TWAP is lower than the market price. The human gatekeepers should ensure that the "market price" at time of request isn't too far from something that is unreasonable (such as a random market fluctuation to the downside). I'd keep this implementation simple for now, as it will likely not be used too frequently.